From e53e10e78f58bf3567aa9de33d814aff3637fd13 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Thu, 7 Mar 2013 15:12:56 -0800
Subject: [PATCH 1/5] qemu-kvm: Add options to pin and prioritize vcpus

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Yang Shi <yang.shi@windriver.com>
Signed-off-by: Paul Barrette <paul.barrette@windriver.com>
Signed-off-by: Michael Barabanov <michael.barabanov@windriver.com>
[Pb: cpus.h in new location for 1.4.2]
[ywei: fix context conflict in vl.c and cpus.h]
Signed-off-by: Yunguo Wei <michael.barabanov@windriver.com>
[Fix context conflict in cpus.c include/sysemu/cpus.h]
Upstream-Status: Pending
Signed-off-by: He Zhe <zhe.he@windriver.com>
---
 cpus.c                | 56 +++++++++++++++++++++++++++++++++++++++++++++++++++
 include/sysemu/cpus.h |  1 +
 qemu-options.hx       |  8 ++++++++
 vl.c                  |  3 +++
 4 files changed, 68 insertions(+)

diff --git a/cpus.c b/cpus.c
index 9bed61ee..ffddb1a0 100644
--- a/cpus.c
+++ b/cpus.c
@@ -1015,6 +1015,60 @@ static void qemu_init_sigbus(void)
 
     prctl(PR_MCE_KILL, PR_MCE_KILL_SET, PR_MCE_KILL_EARLY, 0, 0);
 }
+
+#include <sched.h>
+
+static cpu_set_t vcpu_affinity[1024];
+static int vcpu_priority[1024];
+
+static void qemu_kvm_vcpu_sched_init(CPUState *env)
+{
+    int idx = env->cpu_index;
+
+    if (CPU_COUNT(&vcpu_affinity[idx])) {
+        if (sched_setaffinity(0, sizeof(cpu_set_t), &vcpu_affinity[idx]) < 0)
+             fprintf(stderr,"Setting CPU affinity of VCPU %u to %llx failed with %d\n",
+		     idx, *((unsigned long long *) &vcpu_affinity[idx]), errno);
+    }
+
+    if (vcpu_priority[idx] > 0) {
+        struct sched_param schedp;
+
+        memset(&schedp, 0, sizeof(schedp));
+        schedp.sched_priority = vcpu_priority[idx];
+        if (sched_setscheduler(0, SCHED_FIFO, &schedp) < 0)
+            fprintf(stderr, "Setting priority of VCPU %u to %u failed with %d\n",
+                    idx, vcpu_priority[idx], errno);
+    }
+}
+
+void vcpu_parse(const char *optarg)
+{
+    unsigned long long affinity;
+    int i, cpuid;
+    char option[128];
+    char *endptr;
+
+    cpuid = strtoul(optarg, &endptr, 10);
+    if (endptr != optarg) {
+        if (*endptr == ',') {
+            endptr++;
+        }
+    }
+    if (get_param_value(option, 128, "affinity", endptr) != 0) {
+        affinity = strtoull(option, NULL, 16);
+        for (i = 0; i < 8 * sizeof(unsigned long long); i++) {
+            if (affinity & (1ULL << i))
+                CPU_SET(i, &vcpu_affinity[cpuid]);
+        }
+    }
+    if (get_param_value(option, 128, "prio", endptr) != 0)
+        vcpu_priority[cpuid] = strtoul(option, NULL, 10);
+
+    fprintf(stderr, "vcpu %d: affinity %llx prio %d\n", cpuid,
+	    *((unsigned long long*) &vcpu_affinity[cpuid]), vcpu_priority[cpuid]);
+}
+
 #else /* !CONFIG_LINUX */
 static void qemu_init_sigbus(void)
 {
@@ -1119,6 +1173,8 @@ static void *qemu_kvm_cpu_thread_fn(void *arg)
 
     kvm_init_cpu_signals(cpu);
 
+    qemu_kvm_vcpu_sched_init(cpu);
+
     /* signal CPU creation */
     cpu->created = true;
     qemu_cond_signal(&qemu_cpu_cond);
diff --git a/include/sysemu/cpus.h b/include/sysemu/cpus.h
index 731756d9..a6e000cf 100644
--- a/include/sysemu/cpus.h
+++ b/include/sysemu/cpus.h
@@ -30,6 +30,7 @@ void cpu_synchronize_all_post_init(void);
 void cpu_synchronize_all_pre_loadvm(void);
 
 void qtest_clock_warp(int64_t dest);
+void vcpu_parse(const char *optarg);
 
 #ifndef CONFIG_USER_ONLY
 /* vl.c */
diff --git a/qemu-options.hx b/qemu-options.hx
index 81b0ded9..5c03802f 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -583,6 +583,14 @@ STEXI
 @table @option
 ETEXI
 
+DEF("vcpu", HAS_ARG, QEMU_OPTION_vcpu,
+    "-vcpu n[,affinity=cpumask][,prio=prio] Pin and/or prioritize a vcpu\n", QEMU_ARCH_ALL)
+STEXI
+@item -vcpu @var{opts}
+@findex -vcpu
+Pin vcpu(s) to physical cores and set a realtime priority.
+ETEXI
+
 DEF("fda", HAS_ARG, QEMU_OPTION_fda,
     "-fda/-fdb file  use 'file' as floppy disk 0/1 image\n", QEMU_ARCH_ALL)
 DEF("fdb", HAS_ARG, QEMU_OPTION_fdb, "", QEMU_ARCH_ALL)
diff --git a/vl.c b/vl.c
index 8e247cc2..aced52d1 100644
--- a/vl.c
+++ b/vl.c
@@ -3814,6 +3814,9 @@ int main(int argc, char **argv, char **envp)
                     exit(1);
                 }
                 break;
+            case QEMU_OPTION_vcpu:
+                vcpu_parse(optarg);
+                break;
             case QEMU_OPTION_vnc:
                 vnc_parse(optarg, &error_fatal);
                 break;
-- 
2.13.3

